src/os/bluestore/Writer.cc

```diff
@@ -1023,14 +1037,19 @@ void BlueStore::Writer::_do_put_blobs(
       uint32_t ref_end = std::min(ref_end_offset, data_end_offset);
       //fixme/improve - need something without stupid extras - that is without coll
       left_b->blob->get_ref(coll, in_blob_offset, ref_end - logical_offset);
-      Extent *le = new Extent(
-        logical_offset, in_blob_offset, ref_end - logical_offset, left_b->blob);
-      dout(20) << __func__ << " new extent " << le->print(pp_mode) << dendl;
-      emap.insert(*le);
+      if (left_b->logical_end() == logical_offset) {
+        left_b->length += ref_end - logical_offset;
+        dout(20) << __func__ << " expanded extent " << left_b->print(pp_mode) << dendl;
+      } else {
+        Extent *le = new Extent(
+          logical_offset, in_blob_offset, ref_end - logical_offset, left_b->blob);
+        dout(20) << __func__ << " new extent " << le->print(pp_mode) << dendl;
+        emap.insert(*le);
+      }
+      bstore->logger->inc(l_bluestore_write_small);
+      bstore->logger->inc(l_bluestore_write_small_bytes, ref_end - logical_offset);
```

Comment:
when we check left, we also need check right, for example, there is an extent [4K, 60K], when we write[9K,2K], punch_hole will make extent looks like [4K, 4K] [12K, 52K], we will inset [8K, 4K], we only combined with left make it [4K, 8K], but right one is [12K,52K], we don't combine with right extent.  I think we can reuse extent_map.compress_extent_map  by left_affect_range and right_affect_range that logic is completely and make code readable.